### PR TITLE
federation: Creating kubeconfig files to be used for creating secrets for clusters on aws and gke

### PR DESCRIPTION
**Pull request by nikhiljindal**
Extension of https://github.com/kubernetes/kubernetes/pull/26914 which created the kubeconfig files for gce clusters.
This PR extends it to AWS, vagrant and GKE.

The change for AWS and vagrant is exactly same as GCE.
For GKE, since `gcloud create clusters` creates kubeconfig, we are just copying the generated kubeconfig to the desired location

cc @kubernetes/sig-cluster-federation @colhom 
@roberthbailey for GKE

### CODE REVIEW

**Comment by nikhiljindal** [comment]
Am testing it out

**Comment by fejta** [comment]
@k8s-bot node e2e test this issue #24479

**Comment by nikhiljindal** [comment]
Verified that it works as expected on GKE

File: cluster/gke/util.sh [hunk 1]
```diff
@@ -131,7 +131,6 @@ function validate-cluster {
 #   CLUSTER_IP_RANGE (optional)
 #   GKE_CREATE_FLAGS (optional, space delineated)
 function kube-up() {
-  echo "... in gke:kube-up()" >&2
```
Comment (madhusudancs) [inline cluster/gke/util.sh]: Why remove this?

**Comment by madhusudancs** [comment]
One minor nit. LGTM otherwise. Please feel free to self-apply the label.

File: cluster/gke/util.sh (hunk 1 above)
Comment (nikhiljindal) [inline cluster/gke/util.sh]: Thanks for catching.
reverted

**Comment by nikhiljindal** [comment]
Thanks fixed. Adding LGTM. Also adding P1 since this is for federation which is release blocker for 1.3

**Comment by nikhiljindal** [comment]
Turns out "_" is not allowed in secret names and the context that GKE script creates in kubeconfig is named `gke_myprojectname_us-central1-f_nikhiljindal-gke`. We use the same name to create the secret and hence it fails.

Removing LGTM.

**Comment by nikhiljindal** [comment]
Pushed another commit to fix the context name before using it as cluster and secret names.
Fixing it by replacing "_" by "-" and truncating the string to 253 chars (so that it passes api validation).
Note that this only affects the test setup. For other users, they have to choose the name themselves while running the kubectl create commands.
Have updated E2EContext.Name to be this fixed name which is used everywhere. Added the original name as E2EContext.RawName which no one is using right now, but can be used for debugging.

PTAL.

File: cluster/gke/util.sh [hunk 2]
```diff
@@ -193,6 +193,22 @@ function kube-up() {
...
+  if [[ "${FEDERATION:-}" == "true" ]]; then
+    # Create a kubeconfig with credentials for this apiserver. We will later use
+    # this kubeconfig to create a secret which the federation control plane can
+    # use to talk to this apiserver.
+    KUBECONFIG=${KUBECONFIG:-${HOME}/.kube/config}
+    KUBECONFIG_DIR=$(dirname $KUBECONFIG)
+    CONTEXT=$($KUBECTL config current-context)
+    DEST_KUBECONFIG="${KUBECONFIG_DIR}/federation/kubernetes-apiserver/${CONTEXT}/kubeconfig"
+    mkdir -p $(dirname $DEST_KUBECONFIG) >&2
+    # TODO: Original kubeconfig can contain credential information about
```
Comment (quinton-hoole) [inline cluster/gke/util.sh]: This is pretty dangerous.  Please open an issue and make sure that this TODO gets done soon.

File: cluster/aws/util.sh [hunk 3]
```diff
@@ -1304,7 +1304,18 @@ function build-config() {
...
+   # Update the user's kubeconfig to include credentials for this apiserver.
    create-kubeconfig
+
+   if [[ "${FEDERATION:-}" == "true" ]]; then
+       # Create a kubeconfig with credentials for this apiserver. We will later use
+       # this kubeconfig to create a secret which the federation control plane can
+       # use to talk to this apiserver.
+       KUBECONFIG_DIR=$(dirname ${KUBECONFIG:-$DEFAULT_KUBECONFIG})
+       KUBECONFIG="${KUBECONFIG_DIR}/federation/kubernetes-apiserver/${CONTEXT}/kubeconfig" \
+         create-kubeconfig
```
Comment (quinton-hoole) [inline cluster/aws/util.sh]: nit: probably not worth the line break here.

File: cluster/gke/config-common.sh [hunk 4]
```diff
@@ -26,6 +26,7 @@ NETWORK="${NETWORK:-default}"
 NETWORK_RANGE="${NETWORK_RANGE:-10.240.0.0/16}"
 FIREWALL_SSH="${FIREWALL_SSH:-${NETWORK}-allow-ssh}"
 GCLOUD="${GCLOUD:-gcloud}"
+KUBECTL="${KUBECTL:-kubectl}"
```
Comment (quinton-hoole) [inline cluster/gke/config-common.sh]: nit: It's not totally clear to me why you need this here?

File: cluster/vagrant/util.sh [hunk 5]
```diff
@@ -295,7 +295,17 @@ function kube-up {
    vagrant ssh master -- sudo cat /srv/kubernetes/kubecfg.key >"${KUBE_KEY}" 2>/dev/null
    vagrant ssh master -- sudo cat /srv/kubernetes/ca.crt >"${CA_CERT}" 2>/dev/null
 
+   # Update the user's kubeconfig to include credentials for this apiserver.
    create-kubeconfig
+
+   if [[ "${FEDERATION:-}" == "true" ]]; then
```
Comment (quinton-hoole) [inline cluster/vagrant/util.sh]: nit: Given that this piece of code is duplicated verbatim for 4 (and in general 'n') cloud providers, would it not be better to place it in a common, shared library, and only call that from here?

File: test/e2e/framework/federation_util.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	"fmt"
+	"regexp"
+
+	federationapi "k8s.io/kubernetes/federation/apis/federation"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/validation"
+	validation_util "k8s.io/kubernetes/pkg/util/validation"
+)
+
+// Massages the given name to be a valid dns subdomain name. Most resources (such as secrets, clusters) require the names to be valid dns subdomain.
```
Comment (quinton-hoole) [inline test/e2e/framework/federation_util.go]: nit: Start the doc comment with the function name.  It's the godoc convention.

**Comment by quinton-hoole** [comment]
@nikhiljindal LGTM barring minor nits.  Feel free to self-apply LGTM after addressing them.

I'm guessing that this has not yet been thoroughly tested against AWS and Vagrant, and won't be by the pre-commit tests that will run against this PR.  So I suggest that you keep a close eye on the AWS and Vagrant e2e tests after this PR merges, and confirm that they remain green, or else roll this PR back if not.

**Comment by quinton-hoole** [comment]
Verification failure looks like a transient github or Jenkins failure unrelated to this PR.  While I track down the tracking issue for this place...

@k8s-bot test this issue: IGNORE

**Comment by quinton-hoole** [comment]
```
Checking out Revision fcde5521b2258a6d4bb3b1299ad0373b7287d306 (origin/pr/19485/merge)
 > /usr/bin/git config core.sparsecheckout # timeout=10
 > /usr/bin/git checkout -f fcde5521b2258a6d4bb3b1299ad0373b7287d306 # timeout=20
FATAL: Could not checkout fcde5521b2258a6d4bb3b1299ad0373b7287d306
hudson.plugins.git.GitException: Could not checkout fcde5521b2258a6d4bb3b1299ad0373b7287d306
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$9.execute(CliGitAPIImpl.java:1992)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:152)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:145)
	at hudson.remoting.UserRequest.perform(UserRequest.java:120)
	at hudson.remoting.UserRequest.perform(UserRequest.java:48)
	at hudson.remoting.Request$2.run(Request.java:332)
	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
	at ......remote call to agent-pr-15(Native Method)
	at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1416)
	at hudson.remoting.UserResponse.retrieve(UserRequest.java:220)
	at hudson.remoting.Channel.call(Channel.java:781)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.execute(RemoteGitImpl.java:145)
	at sun.reflect.GeneratedMethodAccessor243.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.invoke(RemoteGitImpl.java:131)
	at com.sun.proxy.$Proxy64.execute(Unknown Source)
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1135)
	at hudson.scm.SCM.checkout(SCM.java:485)
	at hudson.model.AbstractProject.checkout(AbstractProject.java:1269)
	at hudson.model.AbstractBuild$AbstractBuildExecution.defaultCheckout(AbstractBuild.java:607)
	at jenkins.scm.SCMCheckoutStrategy.checkout(SCMCheckoutStrategy.java:86)
	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:529)
	at hudson.model.Run.execute(Run.java:1738)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:410)
Caused by: hudson.plugins.git.GitException: Command "/usr/bin/git checkout -f fcde5521b2258a6d4bb3b1299ad0373b7287d306" returned status code 128:
stdout: 
stderr: fatal: reference is not a tree: fcde5521b2258a6d4bb3b1299ad0373b7287d306

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1719)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$500(CliGitAPIImpl.java:63)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$9.execute(CliGitAPIImpl.java:1984)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:152)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:145)
	at hudson.remoting.UserRequest.perform(UserRequest.java:120)
	at hudson.remoting.UserRequest.perform(UserRequest.java:48)
	at hudson.remoting.Request$2.run(Request.java:332)
	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by quinton-hoole** [comment]
Tracking issue for flake is https://github.com/kubernetes/kubernetes/issues/27505.  Thanks @saad-ali !

**Comment by quinton-hoole** [comment]
```
Jun 15 17:44:59.209: INFO: Couldn't delete ns "e2e-tests-limitrange-jap3s": namespace e2e-tests-limitrange-jap3s was not deleted within limit: timed out waiting for the condition, pods remaining: [pod-no-resources pod-partial-resources]


â€¢ Failure in Spec Teardown (AfterEach) [300.135 seconds]
[k8s.io] LimitRange
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:668
  should create a LimitRange with defaults and ensure pod has those defaults applied. [AfterEach]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/limit_range.go:102

  Jun 15 17:44:59.209: Couldn't delete ns "e2e-tests-limitrange-jap3s": namespace e2e-tests-limitrange-jap3s was not deleted within limit: timed out waiting for the condition, pods remaining: [pod-no-resources pod-partial-resources]
```

**Comment by quinton-hoole** [comment]
Test flake seems unrelated to this PR.  While I find the tracking bug for that flake...

@k8s-bot test this issue: IGNORE

**Comment by quinton-hoole** [comment]
Tracking issue for flake is https://github.com/kubernetes/kubernetes/issues/26290

**Comment by colhom** [comment]
@nikhiljindal the kubeconfig contexts should already be created for you- I don't understand why you are creating additional files in the various `cluster/<provider>/util.sh` files. 

By the time the e2e framework is running `GetUnderlyingFederatedContexts` is all ready to go without any additional changes on to of master.

**Comment by nikhiljindal** [comment]
@colhom You are right. By the time the e2e framework is done, the default kubeconfig is setup with credentials for all the clusters. That enables kubectl to be able to talk to any cluster. That flow is working great.

This change is to be able to create a secret. This secret is used by the federation controller manager (specifically the cluster controller) to be able to talk to each k8s cluster that is part of federation. There is a secret for each k8s cluster. The secret contains the kubeconfig record corresponding to that cluster. For this we create a new kubeconfig file that contains record for only one cluster (vs the default kubeconfig which contains records for all clusters) and then use that newly created kubeconfig file to create the secret.

**Comment by nikhiljindal** [comment]
And yes to be clear, there is another way to do this. Instead of creating those individual kubeconfigs in these shell scripts at cluster bring up time, we can extract those individual records from the single unified kubeconfig (which framework.GetUnderlyingFederatedContexts uses). That way we wont have to write cloud provider specific code and it will be go code so we can have good tests.
But then we will depend on kubeconfig structure and will break if that changes. With the current code, we are immune to that.

Changing the shell script for GCE was easy enough (just 4 lines of code) and the changes were a lot less brittle than the alternative so I went that route.
We can revisit this if maintaining code for all cloud providers turns out too complex.

**Comment by colhom** [comment]
@nikhiljindal the format that comes out of `cluster/kubectl.sh config view` is guaranteed to match what `federation-apiserver` is expecting, as they're using the same resource models and parsing logic?

I don't see the advantage of keeping redundant copies of the kubeconfig around on a per-provider basis. I do see a number of detriments (maintainability, further bloating artifact tree, more potentially non-portable build state, etc)

**Comment by colhom** [comment]
@nikhiljindal and I discussed this offline- I agree that this is the "least bad" way of doing this as of now.

**Comment by colhom** [comment]
LGTM

File: cluster/gke/util.sh (hunk 2 above)
Comment (roberthbailey) [inline cluster/gke/util.sh]: For the GKE specific case, you should be able to use `gcloud container clusters get-credentials` to write a new kubeconfig file with only the cluster info that you care about (not pulling in any existing entries). You can set an env var to stick this in a non-default location. 

/cc @jlowdermilk

**Comment by roberthbailey** [comment]
[sorry for the late comment]

File: cluster/gke/util.sh (hunk 2 above)
Comment (jlowdermilk) [inline cluster/gke/util.sh]: gcloud respects the KUBECONFIG env. So if you want to create a kubeconfig file with a single cluster's info, just
```
export KUBECONFIG=/path/to/standalone/kubeconfig
gcloud container clusters get-credentials mycluster

# do stuff with /path/to/standalone/kubeconfig
```

Comment (jlowdermilk) [inline cluster/gke/util.sh]: BTW, you can create a standalone kubeconfig file for the in-use context using
```
kubectl config view --minify --flatten > /path/to/standalone/kubeconfig
```
The above command is platform agnostic, it's just a yaml dump of the active config.

File: cluster/vagrant/util.sh (hunk 5 above)
Comment (nikhiljindal) [inline cluster/vagrant/util.sh]: done

File: cluster/gke/config-common.sh (hunk 4 above)
Comment (nikhiljindal) [inline cluster/gke/config-common.sh]: not needed anymore. removed

File: cluster/aws/util.sh (hunk 3 above)
Comment (nikhiljindal) [inline cluster/aws/util.sh]: removed

File: cluster/gke/util.sh (hunk 2 above)
Comment (nikhiljindal) [inline cluster/gke/util.sh]: Thanks a lot @jlowdermilk!
Thats an awesome command. Updated the code to use it

**Comment by nikhiljindal** [comment]
Updated the code to use `kubectl config view --minify --flatten` everywhere.
PTAL

**Comment by jlowdermilk** [comment]
lgtm

**Comment by quinton-hoole** [comment]
LGTM.

**Pull request merged**